lib/Doctrine/ORM/Mapping/MetadataCollection.php

```diff
@@ -0,0 +1,47 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\ORM\Mapping;
+
+
+use Doctrine\ORM\Utility\StaticClassNameConverter;
+
+class MetadataCollection
+{
+
+    /**
+     * @var ClassMetadata[]
+     */
+    private $metadata;
+
+    private function __construct(array $metadata)
+    {
+        $metadataByClassName = array_combine(
+            array_map(function (ClassMetadata $metadata) { return $metadata->getClassName(); }, $metadata),
+            $metadata
+        );
+        $this->metadata = $metadataByClassName;
+    }
+
+    public function get(string $name) : ClassMetadata
```

Comment:
The various APIs getting class metadata by object reference could need a `getByObject(object $object)`